### DISCUSSION TITLE
Processing multiple body pairs at an instance

### DISCUSSION THREAD

**Discussion by jenetscaria-mcw** [jrouwe/JoltPhysics#613] [General]
Hi @jrouwe,
While exploring the current pipeline, we saw that according to the current design, after creating a queue of active bodies and finding the corresponding colliding pairs in the queue, we process each pair individually using ProcessBodyPair function and accordingly SSE, AVX2 and AVX512 instructions are used to process individual pair of active bodies. Can we look at processing more body pairs from the queue at once so as to make a complete utilization of 512-bit registers and possibly get a higher performance.

**Comment by jrouwe** [comment]
> Can we look at processing more body pairs from the queue at once so as to make a complete utilization of 512-bit registers and possibly get a higher performance.

Do you mean that *you* want to look into this or are you asking for *me* to look into this?

AVX-512 is not supported on the Intel 12th and 13th gen consumer processors, and while it may be coming back, it means that it will be at least another 10 years before we can reasonably assume that players will have hardware that supports AVX-512. So I don't want to spend much energy in supporting this as the only use case would be games where the server runs the simulation and the client doesn't. My 12th gen Intel doesn't support it, so I cannot test any AVX-512 code myself.

Besides that I think it will be very hard to introduce parallelism at the level of ProcessBodyPair. In that function the two shapes of two bodies are tested for collisions, but the shape types wildly vary from body to body (compounds, spheres, boxes, meshes, heighfields etc.) so if you wanted to run 4 collision tests in parallel using AVX-512 you would have to happen to get 4 shape pairs that have the exact same structure. Let's say that we manage to implement this (maybe as an additional queue) then I think AVX-512 only makes sense for certain collision pairs, e.g. testing a sphere vs sphere is so trivial that the setup cost of moving everything in 512 bit registers probably outweighs the benefit of the faster calculation. It only really makes sense when we're testing complex vs complex shapes (e.g. convex vs mesh). For convex vs mesh the next problem is going to be that the algorithm alternates between testing nodes and testing triangles, so unless the structure of the bounding volume trees for the 4 meshes we're testing in parallel are very similar I don't think we're going to get very good occupancy. The added code complexity for a system like this means that any future changes will become much harder and the code will become very hard to read.

A simpler solution would be to not try to test multiple body pairs at the same time, but to optimize the collision check in e.g. the convex vs mesh case. At the leaves of the bounding volume tree there are X triangles. These triangles all have to be tested against the same shape, so it would be possible to test 4 triangles at the same time. There's no guarantee that this will be much faster though, as testing a convex vs triangle requires an unknown amount of GJK / EPA iterations. In fact, I tried this idea years ago when Jolt was still a test frame work [for ray vs triangle testing](https://github.com/jrouwe/RayCastTest/blob/master/ISPC/RayVsAABBTree.ispc). This version didn't make it into the [final article](http://jrouwe.nl/raycasttest/Rouwe-TriangleEncAndBVHsForRayCasts.pdf) since it was not better than other implementations.

Finally, in my experience it is not the collision checking that is the bottleneck. It is my crappy LockFreeHashMap implementation that dies from cache misses because the key/values belonging to a certain hash are in a linked list. I've been meaning to write a better implementation for this.

**Comment by jrouwe** [comment]
Moving to a discussion as I don't think this is something that can be implemented.

**Comment by jenetscaria-mcw** [comment]
@jrouwe , thanks for the response,
I was just wondering what do you mean by `Cache misses belonging to a certain hash are in a linked list` for the LockFreeHashMap algorithm. From what I saw, in the contact constraint manager, we try to find the cached body pair and if the corresponding key value pair is not found, we create a new entry in the linked list which is inserted as a new element in the beginning of the linked list. 
Can you tell what causes the misses here? Will usage of any other library or data structure be helpful here?

**Comment by jrouwe** [comment]
LockFreeHashMap uses 'Separate Chaining' (see https://en.wikipedia.org/wiki/Hash_table) which means that to look something up in the hash map you first have a cache miss to get the linked list head pointer and then another one to actually get to the first entry (and in case there are multiple entries in the map you'll have more cache misses as you traverse the linked list structure). I think using a different technique like 'Open Addressing' could help but I haven't investigated how that would work for a lock free map.

**Comment by jenetscaria-mcw** [comment]
@jrouwe, if we are able to implement Open Addressing for LockFreeHashMap and with that if we try to process multiple body pairs of similar structure together at once, will it be feasible? Open Addressing can help increase cache performance (maybe by using linear probing), but if we process more body pairs here, will it cause an increase in cache misses?

**Reply by jrouwe** [reply]
I'm missing a lot of context for this discussion. What project are you working on and how are you running into a performance bottleneck? Or is this discussion based on looking at the code and wanting to contribute something to Jolt?

**Reply by jenetscaria-mcw** [reply]
We were actually looking into any scope of vectorization or further optimization so as to contribute to Jolt Physics. That's why, I wanted to check with you whether it would be feasible to vectorize the code in order to process multiple body pairs of similar shapes together.

The batch of active bodies is being fetched in the JobFindCollisions function, so if we are able to process a batch of bodies with similar structures together, that might help to get a good boost in performance. 
Also, for the LockFreeHashMap algorithm, will processing more body pairs lead to increase in cache misses as well?

**Reply by jrouwe** [reply]
> Also, for the LockFreeHashMap algorithm, will processing more body pairs lead to increase in cache misses as well?

No I don't think so. In any case, I would strongly advice against starting vectorization at ProcessBodyPairs level for reasons listed in my first reply. It is unlikely that I would merge this into the master branch as I don't see how it can be done without making the code unreadable/unmaintainable.

**Reply by jenetscaria-mcw** [reply]
@jrouwe , Thanks for the insights,
We are currently checking on Open Addressing method for implementing LockFreeHashMap and simultaneously for optimizing collision check for 4 triangles processing for ConvexVsMesh example.

For the collision check inside the Collide function, after finding the penetration depth, in order to perform the GJK and EPA calculations, will it be fine to modify GetPenetrationDepthStepGJK and GetPenetrationDepthStepEPA functions to process data from all 4 triangles and corresponding convex shapes together?

**Reply by jrouwe** [reply]
I would prefer to keep the current implementation as reference, maybe you can create a new implementation like GJKClosestPoint4. I'm not sure if there will be any benefit in trying to run EPA on 4 triangles at a time because the algorithm only runs in a fraction of the cases. Please make sure to not use any intrinsics in the implementation, intrinsics are only allowed in classes like Vec4, UVec4 etc.

**Reply by jrouwe** [reply]
B.t.w. it will be a lot easier to do a function like MeshShape::sCollideSphereVsMesh first.

**Comment by asenkew** [comment]
Hello Jorrit @jrouwe,
I have question close to this topic, so I decided to write here.
There is hot loop in SolveVelocityConstraints with sSolveVelocityConstraint<EMotionType::Dynamic, EMotionType::Dynamic> call, found in Pyramid test, calculations are vectorized with AVX currently. I believe you have some thoughts about better vectorization here (AVX512 or maybe at least AVX2)? Refactoring could be quite extensive I understand, and I have not realized all details yet. Maybe it was tried already with some result? Anyway, what do you think about it, could you please share your ideas and vision?

**Reply by jrouwe** [reply]
I think it's not easy to refactor the solve velocity loop without reorganizing all data. The code is currently not really setup to allow reorganizing the data. If it will gain any performance, I don't know (it might), but I'm pretty certain that it will leave the code unreadable afterwards. So that's why it hasn't been done and also why it is unlikely to ever happen.

**Reply by asenkew** [reply]
Yes, data reordering is the key point. How strong is this code readability requirement? Can it outweigh performance gain and what is the relation? I believe extended comments could help anyway.

**Reply by jrouwe** [reply]
Code readability and maintainability is pretty important to me. It's hard to put a number to it, but exchanging code for unreadable/unmaintainable code and gaining 10% performance is not worth it. If it became 10x as fast then it would be a different story. My guess though is that it won't be a massive win due to the added costs of data reordering.

I think bigger gains can be made with simpler changes: currently all contacts use the same struct, but if a collision is against a static body or if it is inelastic, a lot of the data in the struct is unused. Packing contacts so they use the minimal amount of memory + templatizing the code to minimize calculations will probably give a nice performance boost (this is on my list to look at).
